maskit/masks.py

```diff
@@ -0,0 +1,296 @@
+            amount is None or amount >= 0
+        ), "Negative values are not supported, please use PerturbationMode.REMOVE"
+        if amount == 0:
+            return
+        count = (
+            abs(amount) if amount is not None else rand.randrange(0, self._mask.size)
+        )
+        if mode == PerturbationMode.ADD:
+            indices = np.argwhere(~self._mask)
+        elif mode == PerturbationMode.INVERT:
+            indices = np.array([list(index) for index in np.ndindex(*self._mask.shape)])
+        elif mode == PerturbationMode.REMOVE:
+            indices = np.argwhere(self._mask)
+        else:
+            raise NotImplementedError(f"The perturbation mode {mode} is not supported")
+        if len(indices) == 0:
+            return
+        indices = tuple(
+            zip(
+                *indices[
+                    np.random.choice(
+                        len(indices), min(count, len(indices)), replace=False
+                    )
+                ]
+            )
+        )
+        self._mask[indices] = ~self._mask[indices]
+
+    def copy(self) -> "MaskedObject":
+        """Returns a copy of the current MaskedObject."""
+        clone = object.__new__(type(self))
+        clone._mask = self._mask.copy()
+        return clone
+
+
+class MaskedParameter(MaskedObject):
+    """
+    A MaskedParameter encapsulates not only the :py:attr:`~.mask` but also the
+    according :py:attr:`~.parameters` being masked.
+    """
+
+    def __init__(self, parameters):
+        super().__init__()
+        self._parameters = parameters
+        self._mask = np.zeros_like(parameters, dtype=bool, requires_grad=False)
+
+    @property
+    def parameters(self):
+        return self._parameters
+
+    @parameters.setter
+    def parameters(self, values):
+        self._parameters = values
+
+    def copy(self) -> "MaskedParameter":
+        clone = super().copy()
+        clone._parameters = self._parameters.copy()
+        return clone
+
+
+class MaskedLayer(MaskedObject):
+    """
+    Logical representation of a MaskedLayer.
+    """
+
+    def __init__(self, layers: int):
+        super().__init__()
+        self._mask = np.zeros(layers, dtype=bool, requires_grad=False)
+
+
+class MaskedWire(MaskedObject):
+    """
+    Logical representation of a MaskedWire.
+    """
+
+    def __init__(self, wires: int):
+        super().__init__()
+        self._mask = np.zeros(wires, dtype=bool, requires_grad=False)
+
+
+class MaskedCircuit(object):
+    """
+    A MaskedCircuit supports masking of different components including wires,
+    layers, and parameters.
+    """
+
+    def __init__(self, parameters: np.ndarray, layers: int, wires: int):
+        assert (
+            layers == parameters.shape[0]
+        ), "First dimension of parameters shape must be equal to number of layers"
+        assert (
+            wires == parameters.shape[1]
+        ), "Second dimension of parameters shape must be equal to number of wires"
```

Comment:
Is there a reason why ``layers``/``wires`` is not automatically taken from the ``parameters.shape``? Specifying all but requiring them to be equal seems redundant.

Reply:
I explicitly want the developer being careful on the selected shape and its meaning for layers and wires. So I currently tend to leave it as is.